### PR TITLE
Fix case where quorum queue consumer is cancelled while message is about to be delivered

### LINKED ISSUE #3729: error on consumer cancel for a quorum queue containing messages
Hello,

I've been encountering weird problems with pika's `basic_cancel` and I've been able to trace it back to quorum queues, specifically when they contain messages. I'm submitting the issue here as I believe quorum queues to be a rabbitmq-specific feature and I do not believe pika to be doing anything fancy in this case.

Minimal python code to replicate the problem:

```
import pika
import time

def dummy():
    pass

def test(queue, arguments, message):
    try:
        print(f"declaring {queue}")
        connection = pika.BlockingConnection(pika.ConnectionParameters(host='localhost'))
        channel = connection.channel()
        channel.queue_declare(queue=queue, durable=True, arguments=arguments)
        if message:
            channel.basic_publish(exchange="", routing_key=queue, body="test")
        ctag = channel.basic_consume(queue, on_message_callback=dummy)

        state = channel.queue_declare(queue=queue, passive=True)
        print(f"current consumers {queue}: {state.method.consumer_count}")

        print(f"cancelling {queue}")
        channel.basic_cancel(ctag)
        time.sleep(2)

        state = channel.queue_declare(queue=queue, passive=True)
        print(f"current consumers {queue}: {state.method.consumer_count}")

        print()
        channel.close()
        connection.close()
    except Exception as e:
        print(f"error: {e}")


test(queue="queue", arguments={}, message=False)

test(queue="quorum_queue", arguments={"x-queue-type": "quorum"}, message=False)

test(queue="queue_with_message", arguments={}, message=True)

test(queue="quorum_queue_with_message", arguments={"x-queue-type": "quorum"}, message=True)
```

When run against rabbitmq 3.8:

```
declaring queue
current consumers queue: 1
cancelling queue
current consumers queue: 0

declaring quorum_queue
current consumers quorum_queue: 1
cancelling quorum_queue
current consumers quorum_queue: 0

declaring queue_with_message
current consumers queue_with_message: 1
cancelling queue_with_message
current consumers queue_with_message: 0

declaring quorum_queue_with_message
current consumers quorum_queue_with_message: 1
cancelling quorum_queue_with_message
current consumers quorum_queue_with_message: 1
```

When run against rabbitmq 3.9:

```
declaring queue
current consumers queue: 1
cancelling queue
current consumers queue: 0

declaring quorum_queue
current consumers quorum_queue: 1
cancelling quorum_queue
current consumers quorum_queue: 0

declaring queue_with_message
current consumers queue_with_message: 1
cancelling queue_with_message
current consumers queue_with_message: 0

declaring quorum_queue_with_message
current consumers quorum_queue_with_message: 1
cancelling quorum_queue_with_message
error: (541, 'INTERNAL_ERROR')
```

[log for the internal error](https://github.com/rabbitmq/rabbitmq-server/files/7539238/rabbitmq.log)

### PR DESCRIPTION
**Pull request by lukebakken**
Fixes #3729

### CODE REVIEW

**Comment by kjnilsson** [comment]
Ok so I'm not sure restoring this behaviour is the right fix for this issue. The problem is that when cancelling a consumer the queue doesn't immediately remove the consumer if there are in-flight messages but the channel side code does remove the consumer from its state.  In an ideal world the two states should be in sync so that the channel can keep delivering any message deliveries that may be queued in the channel mailbox to the consumer. 

In this case a delivery happens and the channel doesn't have the state required. The question is what to do about this. 

An option would be to return the number of in-flight messages that a consumer has on the checkout cancel call to the queue. This way the channel (rabbit_fifo_client) could track the number of messages that are left to processes and remove the consumer state when the last message is acked. This does mean a change to the state machine but it may be possible to do so without threatening machine determinism.

File: deps/rabbit/src/rabbit_fifo_client.erl [hunk 1]
```diff
@@ -738,7 +738,9 @@ handle_delivery(Leader, {delivery, Tag, [{FstId, _} | _] = IdMsgs},
                        consumer_deliveries = CDels0} = State0) ->
     QRef = qref(Leader),
     {LastId, _} = lists:last(IdMsgs),
-    Consumer = #consumer{ack = Ack} = maps:get(Tag, CDels0),
+    %% Note:
+    %% https://github.com/rabbitmq/rabbitmq-server/issues/3729
```
Comment (lhoguin) [inline deps/rabbit/src/rabbit_fifo_client.erl:742]: Please capture the general idea behind the ticket. This URL might not work anymore in 10 years.

**Review by lhoguin — CHANGES_REQUESTED** [review]

File: deps/rabbit/test/quorum_queue_SUITE.erl [hunk 2]
```diff
@@ -2766,6 +2767,43 @@ consumer_priorities(Config) ->
...
+        ct:fail("basic.deliver timeout")
+    end,
+
+    ok = cancel(Ch),
+    ok = timer:sleep(250),
+
+    D = #'queue.declare'{queue = QQ, passive = true, arguments = [{<<"x-queue-type">>, longstr, <<"quorum">>}]},
+    % LRB TODO why is consumer_count still 1 here?
+    ExpectedDeclareRslt1 = #'queue.declare_ok'{queue = QQ, message_count = 0, consumer_count = 1},
+    DeclareRslt1 = amqp_channel:call(Ch, D),
```
Comment (kjnilsson) [inline deps/rabbit/test/quorum_queue_SUITE.erl:2802]: Put a flush in there to see if there are more deliveries at this point

Comment (lukebakken) [reply deps/rabbit/test/quorum_queue_SUITE.erl:2802]: Roger that!

File: deps/rabbit/src/rabbit_fifo_client.erl [hunk 3]
```diff
@@ -732,10 +730,22 @@ maybe_auto_ack(false, {deliver, Tag, _Ack, Msgs} = Deliver, State0) ->
...
 
-
+handle_delivery(_Leader, {delivery, Tag, [_ | _] = IdMsgs},
+                #state{consumer_deliveries = CDels0} = State0) when not is_map_key(Tag, CDels0) ->
+    %% Note:
+    %% https://github.com/rabbitmq/rabbitmq-server/issues/3729
+    %%
+    %% Previously there had been a default value in the consumer deliveries map
+    %% (old code: maps:get(Tag, CDels0, #consumer{last_msg_id = -1}))
+    %% to cover cases where the consumer had been cancelled. We can't restore
```
Comment (kjnilsson) [inline deps/rabbit/src/rabbit_fifo_client.erl:740]: Perhaps we don't need the reference to how it used to work in the bad old days. :)

Let just explain why we're returning here.

File: deps/rabbit/test/quorum_queue_SUITE.erl [hunk 4]
```diff
@@ -2766,6 +2767,55 @@ consumer_priorities(Config) ->
...
+
+    ok = publish(Ch, QQ),
+
+    ok = subscribe(Ch, QQ, false),
+
+    receive
+        {#'basic.deliver'{delivery_tag = DeliveryTag}, _} ->
+            R = #'basic.reject'{delivery_tag = DeliveryTag, requeue = true},
+            ok = amqp_channel:cast(Ch, R)
+    after 1000 ->
```
Comment (kjnilsson) [inline deps/rabbit/test/quorum_queue_SUITE.erl:2791]: I doubt we need to bump it but I probably would for ci anyway.

File: deps/rabbit/src/rabbit_fifo_client.erl [hunk 5]
```diff
@@ -732,10 +730,22 @@ maybe_auto_ack(false, {deliver, Tag, _Ack, Msgs} = Deliver, State0) ->
...
+    %% the default value because the pending messages may not be delivered.
+    %% Instead, we choose to return the messages to the quorum queue process.
+    MsgIntIds = [Id || {Id, _} <- IdMsgs],
+    {State1, Deliveries} = return(Tag, MsgIntIds, State0),
+    {ok, State1, Deliveries};
 handle_delivery(Leader, {delivery, Tag, [{FstId, _} | _] = IdMsgs},
                 #state{cfg = #cfg{cluster_name = QName},
-                       consumer_deliveries = CDels0} = State0) ->
+                       consumer_deliveries = CDels0} = State0) when is_map_key(Tag, CDels0) ->
```
Comment (kjnilsson) [inline deps/rabbit/src/rabbit_fifo_client.erl:748]: As this is the common case I would put it before the other one. Also I do like a newline for the guards when they're wide. :)

**Review by kjnilsson — CHANGES_REQUESTED** [review]
LGMT Minor tweaks only.

File: deps/rabbit/src/rabbit_fifo_client.erl (hunk 3 above)
Comment (lukebakken) [reply deps/rabbit/src/rabbit_fifo_client.erl:740]: haha OK

**Comment by lukebakken** [comment]
@kjnilsson I confirmed that the [repro script](https://github.com/rabbitmq/rabbitmq-server/issues/3729#issue-1053720009) runs correctly using the fixes in this PR.

**Pull request merged**